### PR TITLE
Document Gateway Server MQTT protocol

### LINKED ISSUE #2743: Document Gateway Server MQTT V3 messages
#### Summary

Document Gateway Server MQTT V3 messages

#### Why do we need this ?

For users to know how to interact via MQTT with Gateway Server, using V3-enabled forwarders.

#### What is already there? What do you see now?

The application-layer integration: https://thethingsstack.io/v3.8.4/integrations/mqtt/

#### What is missing? What do you want to see?

Similar documentation for interacting with the Gateway Server via MQTT.

#### How do you propose to document this?

Authentication, message format, topic structure and some example messages for uplink, downlink and status.

#### Can you do this yourself and submit a Pull Request?

Can review

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Closes #2743 

#### Changes

- Add a new section under the Reference page
- Add a link in the Gateway Server page

#### Notes for reviewers

- @benolayinka Formatting, presentation
- @KrishnaIyer @johanstokking content, scope

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [X] Documentation: Relevant documentation is added or updated.
- [X] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: doc/content/reference/gs-mqtt-v3-protocol/_index.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+title: "Gateway Server MQTT v3 Protocol"
```
Comment (johanstokking) [inline doc/content/reference/gs-mqtt-v3-protocol/_index.md]: This stack version is V3; no need to use in documentation folders and titles

File: doc/content/reference/gs-mqtt-v3-protocol/downlink.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+---
+title: "Downlink messages"
+description: ""
+weight: 30
+---
+
+The Gateway Server instructs the gateway to send a downlink packet by publishing a protocol buffer of type `ttnpb.GatewayDown` under the topic `v3/<gateway-id>/down`.
```
Comment (johanstokking) [inline doc/content/reference/gs-mqtt-v3-protocol/downlink.md]: Protocol Buffers is a name (title case)

File: doc/content/reference/gs-mqtt-v3-protocol/status.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
+---
+title: "Status messages"
+description: ""
+weight: 20
+---
+
+To forward a gateway status message to the Gateway Server, the MQTT client must publish a marshalled protocol buffer of type `ttnpb.GatewayStatus` under the topic `v3/<gateway-id>/status`.
```
Comment (johanstokking) [inline doc/content/reference/gs-mqtt-v3-protocol/status.md]: Idem

**Review by johanstokking — COMMENTED** [review]
Good stuff.

I think the big question for consumers is how to generate the protos so they can marshal and unmarshal.

@rvolosatovs what do you think about this moving forward? I think it's quite challenging today with all the options. Few solutions;

- target C/C++ with our protoc and put in a separate repo
- extract gateway client relevant protos and maintain manually in a separate repo
- probably nicest: generate clean protos from our protos that can be used to generate with vanilla protoc to any platform

File: doc/content/reference/gs-mqtt-protocol/_index.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+---
+title: "Gateway Server MQTT Protocol"
+description: "Gateway Server MQTT Protocol"
```
Comment (johanstokking) [inline doc/content/reference/gs-mqtt-protocol/_index.md]: Browsing in the local doc server, I'd find this more appropriate under API, and that it shows up as "Gateway Server MQTT"

File: doc/content/reference/gs-mqtt-protocol/status.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+```bash
+$ export GATEWAY_ID="test-gtw"
+$ export GATEWAY_API_KEY="NNSXS.VEEBURF3KR77ZR..." # API key with RIGHT_GATEWAY_LINK rights
+$ mosquitto_pub \
+    -h "thethings.example.com" -p 1883 \
+    -u "$GATEWAY_ID" -P "$GATEWAY_API_KEY" \
+    -t "v3/$GATEWAY_ID/status" -f test-gateway-status
+```
+
+>Note: The file `test-gateway-status` contains the raw bytes of the Protocol Buffer, not a JSON object.
```
Comment (johanstokking) [inline doc/content/reference/gs-mqtt-protocol/status.md]: Hmm, maybe "contains a Protocol Buffers file (binary, not JSON)." ?

**Review by johanstokking — COMMENTED** [review]
Almost there.

@rvolosatovs still input needed on how to generate this, see above

File: doc/content/reference/gs-mqtt-protocol/_index.md (hunk 4 above)
Comment (neoaggelos) [reply doc/content/reference/gs-mqtt-protocol/_index.md]: Given that this not a "user-facing" service, does it make sense for it to appear next to the gRPC and HTTP APIs?

EDIT: By user-facing, I mean that end-users should not have to use this under normal circumstances.

Comment (johanstokking) [reply doc/content/reference/gs-mqtt-protocol/_index.md]: Define end-users. The target audience for the API section is developers, and those can be of all sorts: packet forwarder developers and app developers to manage applications in TTES.

Comment (neoaggelos) [reply doc/content/reference/gs-mqtt-protocol/_index.md]: Alright

File: doc/content/reference/api/gateway_server_mqtt.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+Clients also get **read-only** access and should subscribe to the following topic:
+
+- `v3/<gateway-id>/down`: The Gateway Server publishes downlink message that the gateway should transmit.
+
+## Disconnect from the Gateway Server
+
+The gateway can disconnect by terminating the MQTT client connection.
+
+## Uplink messages
```
Comment (johanstokking) [inline doc/content/reference/api/gateway_server_mqtt.md]: Title case, also other occurrences

File: doc/content/reference/api/gateway_server_mqtt.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+```bash
+$ export GATEWAY_ID="test-gtw"
+$ export GATEWAY_API_KEY="NNSXS.VEEBURF3KR77ZR..." # API key with RIGHT_GATEWAY_LINK rights
+$ mosquitto_pub \
+    -h "thethings.example.com" -p 1883 \
+    -u "$GATEWAY_ID" -P "$GATEWAY_API_KEY" \
+    -t "v3/$GATEWAY_ID/up" -f test-uplink-message
+```
+
+>Note: The file `test-uplink-message` contains a Protocol Buffer of type `ttnpb.UplinkMessage` (binary, not JSON).
```
Comment (johanstokking) [inline doc/content/reference/api/gateway_server_mqtt.md]: NOTE:

File: doc/content/reference/api/gateway_server_mqtt.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+To forward uplink traffic to the Gateway Server, the MQTT client must publish a Protocol Buffer of type `ttnpb.UplinkMessage` under the topic `v3/<gateway-id>/up`.
+
+Below is an example that connects to the Gateway Server as `$GATEWAY_ID`, publishes an uplink message (Protocol Buffer stored as binary file `test-uplink-message`) and disconnects:
+
+```bash
+$ export GATEWAY_ID="test-gtw"
+$ export GATEWAY_API_KEY="NNSXS.VEEBURF3KR77ZR..." # API key with RIGHT_GATEWAY_LINK rights
+$ mosquitto_pub \
+    -h "thethings.example.com" -p 1883 \
```
Comment (johanstokking) [inline doc/content/reference/api/gateway_server_mqtt.md]: Probably note that this is insecure

File: doc/content/reference/api/gateway_server_mqtt.md [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+The gateway can disconnect by terminating the MQTT client connection.
+
+## Uplink messages
+
+To forward uplink traffic to the Gateway Server, the MQTT client must publish a Protocol Buffer of type `ttnpb.UplinkMessage` under the topic `v3/<gateway-id>/up`.
+
+Below is an example that connects to the Gateway Server as `$GATEWAY_ID`, publishes an uplink message (Protocol Buffer stored as binary file `test-uplink-message`) and disconnects:
+
+```bash
+$ export GATEWAY_ID="test-gtw"
```
Comment (johanstokking) [inline doc/content/reference/api/gateway_server_mqtt.md:47]: Be sure to update this in TTES as well with tenant ID

**Review by johanstokking — APPROVED** [review]

**Comment by rvolosatovs** [comment]
> @rvolosatovs what do you think about this moving forward? I think it's quite challenging today with all the options. Few solutions;
> 
>     * target C/C++ with our protoc and put in a separate repo
> 
>     * extract gateway client relevant protos and maintain manually in a separate repo
> 
>     * probably nicest: generate clean protos from our protos that can be used to generate with vanilla protoc to any platform

I think nicest solution would be to distinguish between "internal" protos and "external" protos, where in the "internal" protos we would use whatever custom options we would want to use and we would keep the "external" protos clean.
We could then have generated protos contained in SDKs. And if SDK is missing, users could generate the "external" protos.

I don't know if that's feasible to do this separation manually at this point, but we could try. Perhaps a better approach would be to write a tool, which would take a list of names as parameter and then extract those proto entities into "vanilla" protos.

So, essentially, I think we should do a mix of all 3:
- Generate clean protos containing only external API - I would prefer to have (from repo root) `internal/api` and `api`, where we move all our current protos to `internal/api` and then generate clean, vanilla protos (`.proto` files) in `/api` (perhaps including any possible proprietary APIs as well)
- Contain generated protos in SDKs (generated protos are then the starting point of an SDK - so for a language X, an MVP of a X-SDK is just a collection of generated "clean" external protos in language X)

**Comment by neoaggelos** [comment]
> > @rvolosatovs what do you think about this moving forward? I think it's quite challenging today with all the options. Few solutions;
> > ```
> > * target C/C++ with our protoc and put in a separate repo
> > 
> > * extract gateway client relevant protos and maintain manually in a separate repo
> > 
> > * probably nicest: generate clean protos from our protos that can be used to generate with vanilla protoc to any platform
> > ```
> 
> I think nicest solution would be to distinguish between "internal" protos and "external" protos, where in the "internal" protos we would use whatever custom options we would want to use and we would keep the "external" protos clean.
> We could then have generated protos contained in SDKs. And if SDK is missing, users could generate the "external" protos.
> 
> I don't know if that's feasible to do this separation manually at this point, but we could try. Perhaps a better approach would be to write a tool, which would take a list of names as parameter and then extract those proto entities into "vanilla" protos.
> 
> So, essentially, I think we should do a mix of all 3:
> 
> * Generate clean protos containing only external API - I would prefer to have (from repo root) `internal/api` and `api`, where we move all our current protos to `internal/api` and then generate clean, vanilla protos (`.proto` files) in `/api` (perhaps including any possible proprietary APIs as well)
> * Contain generated protos in SDKs (generated protos are then the starting point of an SDK - so for a language X, an MVP of a X-SDK is just a collection of generated "clean" external protos in language X)

I suppose this is out of scope for this PR, right? The discussion should be in a separate issue.

**Comment by johanstokking** [comment]
> So, essentially, I think we should do a mix of all 3:
> 
> - Generate clean protos containing only external API - I would prefer to have (from repo root) `internal/api` and `api`, where we move all our current protos to `internal/api` and then generate clean, vanilla protos (`.proto` files) in `/api` (perhaps including any possible proprietary APIs as well)
> 
> - Contain generated protos in SDKs (generated protos are then the starting point of an SDK - so for a language X, an MVP of a X-SDK is just a collection of generated "clean" external protos in language X)

I think this is a great idea.

@neoaggelos yes this is out of scope for this PR. But please file an issue with the conclusion here, still labeled `discussion` and loop in @htdvisser there too.

**Review by benolayinka — APPROVED** [review]
Looks great

**Pull request merged**